File: todo.py
import json
import os
import sys

DATA_FILE = 'tasks.json'

def load_tasks():
    if os.path.exists(DATA_FILE):
        with open(DATA_FILE, 'r', encoding='utf-8') as f:
            return json.load(f)
    return []

def save_tasks(tasks):
    with open(DATA_FILE, 'w', encoding='utf-8') as f:
        json.dump(tasks, f, ensure_ascii=False, indent=2)

def list_tasks():
    tasks = load_tasks()
    if not tasks:
        print('No tasks.')
    else:
        for idx, t in enumerate(tasks, 1):
            status = '✓' if t.get('done') else ' '
            print(f"{idx}. [{status}] {t['text']}")

def add_task(text):
    tasks = load_tasks()
    tasks.append({'text': text, 'done': False})
    save_tasks(tasks)
    print('Task added.')

def complete_task(index):
    tasks = load_tasks()
    if 0 <= index < len(tasks):
        tasks[index]['done'] = True
        save_tasks(tasks)
        print('Task completed.')
    else:
        print('Invalid task number.')

def delete_task(index):
    tasks = load_tasks()
    if 0 <= index < len(tasks):
        tasks.pop(index)
        save_tasks(tasks)
        print('Task deleted.')
    else:
        print('Invalid task number.')

def usage():
    print("Usage: python todo.py [list|add TEXT|done NUMBER|delete NUMBER]")

if __name__ == '__main__':
    if len(sys.argv) < 2:
        usage()
    else:
        cmd = sys.argv[1]
        if cmd == 'list':
            list_tasks()
        elif cmd == 'add':
            if len(sys.argv) < 3:
                print('Provide task text.')
            else:
                add_task(' '.join(sys.argv[2:]))
        elif cmd == 'done':
            if len(sys.argv) < 3:
                print('Provide task number.')
            else:
                try:
                    index = int(sys.argv[2]) - 1
                    complete_task(index)
                except ValueError:
                    print('Provide a number.')
        elif cmd == 'delete':
            if len(sys.argv) < 3:
                print('Provide task number.')
            else:
                try:
                    index = int(sys.argv[2]) - 1
                    delete_task(index)
                except ValueError:
                    print('Provide a number.')
        else:
            usage()

File: Todo.java
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.*;
import java.util.*;
import java.util.regex.*;

public class Todo {
    private static final String DATA_FILE = "tasks.json";

    static class Task {
        String text;
        boolean done;

        Task(String text, boolean done) {
            this.text = text;
            this.done = done;
        }
    }

    private static String escape(String s) {
        return s.replace("\\", "\\\\").replace("\"", "\\\"");
    }

    private static List<Task> loadTasks() {
        List<Task> tasks = new ArrayList<>();
        Path p = Paths.get(DATA_FILE);
        if (!Files.exists(p)) {
            return tasks;
        }
        try {
            String content = new String(Files.readAllBytes(p), StandardCharsets.UTF_8);
            Pattern pattern = Pattern.compile("\\{\\s*\"text\"\\s*:\\s*\"(.*?)\"\\s*,\\s*\"done\"\\s*:\\s*(true|false)\\s*\\}");
            Matcher m = pattern.matcher(content);
            while (m.find()) {
                String text = m.group(1);
                boolean done = Boolean.parseBoolean(m.group(2));
                tasks.add(new Task(text, done));
            }
        } catch (IOException e) {
            System.err.println("Failed to load tasks: " + e.getMessage());
        }
        return tasks;
    }

    private static void saveTasks(List<Task> tasks) {
        StringBuilder sb = new StringBuilder();
        sb.append("[\n");
        for (int i = 0; i < tasks.size(); i++) {
            Task t = tasks.get(i);
            sb.append("  {\"text\":\"").append(escape(t.text)).append("\",\"done\":").append(t.done);
            sb.append("}");
            if (i < tasks.size() - 1) sb.append(",");
            sb.append("\n");
        }
        sb.append("]\n");
        try {
            Files.write(Paths.get(DATA_FILE), sb.toString().getBytes(StandardCharsets.UTF_8));
        } catch (IOException e) {
            System.err.println("Failed to save tasks: " + e.getMessage());
        }
    }

    private static void listTasks() {
        List<Task> tasks = loadTasks();
        if (tasks.isEmpty()) {
            System.out.println("No tasks.");
        } else {
            for (int i = 0; i < tasks.size(); i++) {
                Task t = tasks.get(i);
                String status = t.done ? "\u2713" : " ";
                System.out.printf("%d. [%s] %s%n", i + 1, status, t.text);
            }
        }
    }

    private static void addTask(String text) {
        List<Task> tasks = loadTasks();
        tasks.add(new Task(text, false));
        saveTasks(tasks);
        System.out.println("Task added.");
    }

    private static void completeTask(int index) {
        List<Task> tasks = loadTasks();
        if (index >= 0 && index < tasks.size()) {
            tasks.get(index).done = true;
            saveTasks(tasks);
            System.out.println("Task completed.");
        } else {
            System.out.println("Invalid task number.");
        }
    }

    private static void deleteTask(int index) {
        List<Task> tasks = loadTasks();
        if (index >= 0 && index < tasks.size()) {
            tasks.remove(index);
            saveTasks(tasks);
            System.out.println("Task deleted.");
        } else {
            System.out.println("Invalid task number.");
        }
    }

    private static void usage() {
        System.out.println("Usage: java Todo [list|add TEXT|done NUMBER|delete NUMBER]");
    }

    public static void main(String[] args) {
        if (args.length < 1) {
            usage();
            return;
        }
        String cmd = args[0];
        switch (cmd) {
            case "list":
                listTasks();
                break;
            case "add":
                if (args.length < 2) {
                    System.out.println("Provide task text.");
                } else {
                    String text = String.join(" ", Arrays.copyOfRange(args, 1, args.length));
                    addTask(text);
                }
                break;
            case "done":
                if (args.length < 2) {
                    System.out.println("Provide task number.");
                } else {
                    try {
                        int index = Integer.parseInt(args[1]) - 1;
                        completeTask(index);
                    } catch (NumberFormatException e) {
                        System.out.println("Provide a number.");
                    }
                }
                break;
            case "delete":
                if (args.length < 2) {
                    System.out.println("Provide task number.");
                } else {
                    try {
                        int index = Integer.parseInt(args[1]) - 1;
                        deleteTask(index);
                    } catch (NumberFormatException e) {
                        System.out.println("Provide a number.");
                    }
                }
                break;
            default:
                usage();
        }
    }
}
